### PR TITLE
feat: sharing buckets

### PR DESCRIPTION
**Pull request by benciks**

### CODE REVIEW

File: pkg/graph/drive.resolvers.go [hunk 1]
```diff
@@ -424,6 +457,142 @@ func (r *mutationResolver) CreateSharedDrive(ctx context.Context, name string) (
...
+		Permission:     strings.ToLower(string(input.Permission)),
+	}
+
+	if input.FileID != nil {
+		share.FileID = *input.FileID
+	} else {
+		share.BucketID = *input.BucketID
+	}
+
+	err = r.DB.NewInsert().Model(&share).Returning("*").Scan(ctx)
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go:548]: Are we before inserting checking if the user has the permission to share this folder/bucket?

File: pkg/graph/drive.resolvers.go [hunk 2]
```diff
@@ -424,6 +457,142 @@ func (r *mutationResolver) CreateSharedDrive(ctx context.Context, name string) (
...
+			Scan(ctx)
+	} else if input.BucketID != nil {
+		err = r.DB.NewSelect().Model(&share).
+			Where("shared_with= ?", input.User).
+			Where("organisation_id = ?", currentUser.OrganisationID).
+			Where("bucket_id = ?", input.BucketID).
+			Scan(ctx)
+	} else {
+		return nil, errors.New("fileId or bucketId is required")
+	}
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go]:
```suggestion
	query := r.DB.NewSelect().
		Model(&share).
		Where("shared_with= ?", input.User).
		Where("organisation_id = ?", currentUser.OrganisationID)

	if input.FileID != nil {
		query = query.Where("file_id = ?", input.FileID)
	} else if input.BucketID != nil {
		query = query.Where("bucket_id = ?", input.BucketID)
	} else {
		return nil, errors.New("fileId or bucketId is required")
	}

	err = query.Scan(ctx)
```

File: pkg/graph/drive.resolvers.go [hunk 3]
```diff
@@ -424,6 +457,142 @@ func (r *mutationResolver) CreateSharedDrive(ctx context.Context, name string) (
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Update permission
+	share.Permission = strings.ToLower(string(input.Permission))
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go:639]: Is there a cleaner way to handle the parsing of `input.Permission`? Perhaps type the input?

File: pkg/graph/drive.resolvers.go [hunk 4]
```diff
@@ -424,6 +457,142 @@ func (r *mutationResolver) CreateSharedDrive(ctx context.Context, name string) (
...
+	err = r.DB.NewUpdate().Model(&share).WherePK().Scan(ctx)
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go]: Missing my beautiful `Where("organisation_id = ?")`... _(also in a few other places)_

File: pkg/graph/drive.resolvers.go [hunk 5]
```diff
@@ -424,6 +457,142 @@ func (r *mutationResolver) CreateSharedDrive(ctx context.Context, name string) (
...
+		Where("shared_with = ?", input.User).
+		Where("organisation_id = ?", currentUser.OrganisationID)
+
+	if input.FileID != nil {
+		query.Where("file_id = ?", input.FileID)
+	} else if input.BucketID != nil {
+		query.Where("bucket_id = ?", input.BucketID)
+	} else {
+		return nil, errors.New("fileId or bucketId is required")
+	}
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go:683]: oh, nice! this already has the query outside the if statement

File: pkg/graph/drive.resolvers.go [hunk 6]
```diff
@@ -424,6 +457,142 @@ func (r *mutationResolver) CreateSharedDrive(ctx context.Context, name string) (
...
+
+	err = query.Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Delete share
+	_, err = r.DB.NewDelete().Model(&share).WherePK().Exec(ctx)
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go]: Before doing that we should check if the user is allowed to delete the share

File: pkg/graph/drive.resolvers.go [hunk 7]
```diff
@@ -462,6 +631,39 @@ func (r *queryResolver) Buckets(ctx context.Context, input *model.BucketFilterIn
...
+			Where("shared = ?", *input.Shared).
+			Where("id IN (?)", bun.In(bucketIDs)).
+			Scan(ctx)
+		if err != nil {
+			if err != sql.ErrNoRows {
+				return nil, err
+			}
+		}
+
+		buckets = append(buckets, sharedBuckets...)
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go:833]: I think we should do this beforehand and have it within `one`-sql query. More performant and perhaps more secure

File: pkg/graph/drive.resolvers.go [hunk 8]
```diff
@@ -501,7 +703,23 @@ func (r *queryResolver) Bucket(ctx context.Context, id string) (*db.Bucket, erro
...
+			// Fetch the bucket, if share exists
+			if &share != nil {
+				err = r.DB.NewSelect().Model(&bucket).Where("id = ?", share.BucketID).Scan(ctx)
+				if err != nil {
+					return nil, err
+				}
+			}
+		} else {
+			return nil, err
+		}
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go]: This also can be cleaner. Perhaps part of the first sql query

File: pkg/graph/drive.resolvers.go [hunk 9]
```diff
@@ -619,6 +837,59 @@ func (r *queryResolver) Files(ctx context.Context, input *model.FilesFilterInput
...
+
+	var shares []*db.Share
+
+	if input.FileID != nil {
+		err = r.DB.NewSelect().Model(&shares).Where("file_id = ?", input.FileID).Scan(ctx)
+	} else if input.BucketID != nil {
+		err = r.DB.NewSelect().Model(&shares).Where("bucket_id = ?", input.BucketID).Scan(ctx)
+	} else {
+		return nil, fmt.Errorf("file_id or bucket_id must be provided")
+	}
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go]: move query outside if statement

File: pkg/graph/drive.resolvers.go [hunk 10]
```diff
@@ -619,6 +837,59 @@ func (r *queryResolver) Files(ctx context.Context, input *model.FilesFilterInput
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Make slice of user IDs
+	var userIDs []string
+	for _, share := range shares {
+		userIDs = append(userIDs, share.SharedWith)
+	}
```
Comment (tomhaerter) [inline pkg/graph/drive.resolvers.go:1053]: I don't think we need to do that. Having this within a sql query should be more secure and faster

**Review by tomhaerter — CHANGES_REQUESTED** [review]
This is a great first implementation! It shows what and how to do it and I am sure we can resolve some of the rough edges together. Please let me know in case you'd like some ideas for or help with the SQL queries

File: pkg/graph/drive.resolvers.go (hunk 7 above)
Comment (benciks) [reply pkg/graph/drive.resolvers.go:833]: How is it possible to put it in one query, where in one you clearly fetch by user_id in bucket, which won't fetch any shared buckets?

File: pkg/graph/drive.resolvers.go (hunk 8 above)
Comment (benciks) [reply pkg/graph/drive.resolvers.go]: What does this mean?

**Pull request merged**